tools/MultiProjectValidator/AnalysisRules/DependencyMismatchRule.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.DotNet.ProjectModel;
+
+namespace MultiProjectValidator.AnalysisRules
+{
+    public class DependencyMismatchRule : IAnalysisRule
+    {
+        public AnalysisResult Evaluate(List<ProjectContext> projectContexts)
+        {
+            var targetGroupedContexts = GroupContextsByTarget(projectContexts);
+
+            var failureMessages = EvaluateTargetContextGroups(targetGroupedContexts);
+            var pass = failureMessages.Count == 0;
+
+            var result = new AnalysisResult(failureMessages, pass);
+            return result;
+        }
+
+        private List<string> EvaluateTargetContextGroups(Dictionary<string, List<ProjectContext>> targetGroupedContexts)
+        {
+            var failureMessages = new List<string>();
+
+            foreach (var target in targetGroupedContexts.Keys)
+            {
+                var targetContexts = targetGroupedContexts[target];
+
+                failureMessages.AddRange(EvaluateTargetContextGroup(targetContexts));
+            }
+
+            return failureMessages;
+        }
+
+        private List<string> EvaluateTargetContextGroup(List<ProjectContext> targetContexts)
+        {
+            var failedMessages = new List<string>();
+            var assemblyVersionMap = new Dictionary<string, string>();
+
+            foreach(var context in targetContexts)
+            {
+                var libraries = context.LibraryManager.GetLibraries();
+
+                foreach(var library in libraries)
+                {
+                    var name = library.Identity.Name;
+                    var version = library.Identity.Version.ToString();
+
+                    if (assemblyVersionMap.ContainsKey(name))
+                    {
+                        var existingVersion = assemblyVersionMap[name];
+                        if (!string.Equals(existingVersion, version, StringComparison.OrdinalIgnoreCase))
+                        {
+                            string message = 
```

Comment:
Suppose there are three projects that reference LibA
P1 -> LibA 1.0
P2 -> LibA 2.0
P3 -> LibA 2.0

As I understand this logic, if P1 is the first project analyzed, it won't trigger an analysis failure when analyzed.  P2 and P3 will however and their error messages will look like the following - "Dependency mismatch in {Path for P2/P3} for dependency LibA.  Versions 2.0, 1.0."  In this scenario both P2 and P3 depend on APIs that don't exist in LibA 1.0 therefore the P1 reference is what needs to be updated since it is an older version.  What is unfortunate about this scenario is that there isn't a failure message for the project that needs to change - P1.  The logic assume the first project found is correct.

Would it be better to do one of the following?
1. Whenever a mismatch is found emit a failure message for each project that references the dependency.
2. Whenever a mismatch is found emit a single failure message that aggregates all of the information such as the following.

Dependency mismatch for 'LibA'.
    Path1/P1 (1.0)
    Path2/P2 (2.0)
    Path3/P3 (2.0)